### ISSUE TITLE
Appium values for element vary from console (actual) values. Have inserted 0 on iOS only. 

### ISSUE THREAD

**Issue by StephanLuis** [appium/appium#16697]
## The problem

I'm getting values from a time input plugin control that have an extra 0 using appium for iOS.

 So "value":"00:007:00.000" ! Notice the 007 portion for minutes. 

The console value is 00:07:00.000 on iOS with webinspector, and in iOS netscape console on Browserstack.  Other OS/Browser combinations are fine. 

## Environment
Appium v1.17.0 creating new XCUITestDriver (v3.21.1) session

Using Broswerstack I've experienced this bug for Iphone12, Iphone SE 2020, and iPads, so presuming an appium/ iOS bug.  I've verified the correct value on real iPad using webinspector. 

## Details

If necessary, describe the problem you have been experiencing in more detail.

## Link to Appium logs
[https://gist.github.com/StephanLuis/5c4f2259a61fd7f0d443190db74bb52e](full Appium Log)

Offending Line:

2022-03-30 10:53:25:498 - [debug] [RemoteDebugger] Received data response from send (id: 70): '{"status":0,"value":"00:007:00.000"}'

## Code To Reproduce Issue [ Good To Have ]

Node Code:

await driver.findElement(By.xpath('//*[@id="bob"]/input[2]')).click();
await driver.findElement(By.xpath('//*[@id="bob"]/input[2]')).sendKeys("7");

**Comment by jlipps** [comment]
What happens if you run the same javascript that Appium does, in the webinspector, at the appropriate part in the test:

```js
return document.querySelector("#bob").value
```

What does it return?

**Comment by StephanLuis** [comment]
I get the correct value.  Tested on iOS, iPad.  Quirky part is that Appium returns the correct value from other browsers:  laptop (osX and Windows) and from Android devices.

**Comment by jlipps** [comment]
Very strange. In our case it would appear then that the issue would be in the built selenium atoms code, which is the bundle of JS we send to mimic webdriver commands. We don't implement the code for these atoms ourselves, rather we just build them from the selenium project. It's possible we could see if there is a new version that's possible to build to see if it fixes the issue. what do you think @mykola-mokhnach ?

**Comment by mykola-mokhnach** [comment]
No idea, @jlipps 🤷 
I don't have much expertise with remote debugger stuff

**Comment by StephanLuis** [comment]
Would it be possible for appium to open an issue with them?  I don't know anything about the selenium atoms/ modules nor how to determine which is affected.  My time control is a built of four input controls; what I can do is write a more specific selenium test so that only the erroneous value is returned. That may help isolate the bug.  But that would need to be passed on to the correct selenium guys.

**Comment by jlipps** [comment]
@AutomatedTester have you seen anything like this weird response from the selenium atoms inserting an extra '0' in a datetime value? Are the atoms maintained these days?

**Comment by AutomatedTester** [comment]
We definitely do maintain that code, though its not needed maintenance for some time.

I haven't seen an issue like this before. Is this coming from the `getAttribute` atom?

**Comment by jlipps** [comment]
In this case it's coming from the execute script atom. Here is Appium's logging for the situation:

```
2022-03-30 10:53:25:473 - [HTTP] --> POST /wd/hub/session/f7190d3f-2759-4c2d-8d53-9bd3aa0bc270/execute/sync
2022-03-30 10:53:25:473 - [HTTP] {"script":"return document.querySelector(\"#bob\").value","args":[]}
2022-03-30 10:53:25:473 - [debug] [W3C (f7190d3f)] Calling AppiumDriver.execute() with args: ["return document.querySelector(\"#bob\").value",[],"f7190d3f-2759-4c2d-8d53-9bd3aa0bc270"]
2022-03-30 10:53:25:473 - [debug] [XCUITest] Executing command 'execute'
2022-03-30 10:53:25:474 - [debug] [RemoteDebugger] Executing atom 'execute_script'
2022-03-30 10:53:25:474 - [debug] [RemoteDebugger] Executing 'execute_script' atom in default context
2022-03-30 10:53:25:474 - [debug] [RemoteDebugger] Sending javascript command: '(function(){return function(){var f=this;
2022-03-30 10:53:25:474 - [debug] [RemoteDebugger] funct...'
2022-03-30 10:53:25:474 - [debug] [RemoteDebugger] Sending '_rpc_forwardSocketData:' message to app 'PID:5869', page '1', target 'page-13' (id: 70): 'Runtime.evaluate'
2022-03-30 10:53:25:498 - [debug] [RemoteDebugger] Received data response from send (id: 70): '{"status":0,"value":"00:007:00.000"}'
2022-03-30 10:53:25:498 - [debug] [RemoteDebugger] Sending to Web Inspector took 24ms
2022-03-30 10:53:25:498 - [debug] [RemoteDebugger] Received result for atom 'execute_script' execution: "00:007:00.000"
2022-03-30 10:53:25:505 - [debug] [W3C (f7190d3f)] Responding to client with driver.execute() result: "00:007:00.000"
2022-03-30 10:53:25:506 - [HTTP] <-- POST /wd/hub/session/f7190d3f-2759-4c2d-8d53-9bd3aa0bc270/execute/sync 200 33 ms - 25
```

but @StephanLuis I just realized you're using a quite old version of Appium and the XCUITest driver. Can you try again with Appium 2 + the latest XCUITest driver? The remote debugger code has seen some updates since then including additional logging.

**Comment by StephanLuis** [comment]
@jlipps I'm running this on Browserstack, is there another (free preferably) webservice I could run this on with up to date Appium?  I don't have Selenium/ Appium installed on a computer/ gitpod container. Thx.

**Comment by StephanLuis** [comment]
Another thought that I want to post ... my code has an event listener that adds 0s.  If the iOS driver emits additional events (that are not intended) that could potentially be a source of this behaviour.  Remember, Appium is correct for all other OS than iOS.

**Comment by jlipps** [comment]
> I'm running this on Browserstack, is there another (free preferably) webservice I could run this on with up to date Appium? I don't have Selenium/ Appium installed on a computer/ gitpod container. Thx.

We'll definitely want to run on latest to investigate this any further. I would recommend configuring Appium locally to check.

> Another thought that I want to post ... my code has an event listener that adds 0s.

Well this is almost certainly the cause then. What event listener is hooked up?

>  If the iOS driver emits additional events (that are not intended) that could potentially be a source of this behaviour

The Selenium atoms work by triggering events on elements, so yes. Maybe some previous step has inadvertently triggered an event?

> Remember, Appium is correct for all other OS than iOS.

It is only on iOS that Appium uses the Selenium atoms, so that would not be surprising. For Android it uses Chromedriver which is a completely different set of automation technologies.

**Comment by StephanLuis** [comment]
@jlipps  My add0s() is driven by the 'change' event on each input (numeric).

I had a quick look at the Appium getting started guides, have you seen a demo Github repo that I could run on Gitpod?

**Comment by jlipps** [comment]
the current getting started guides and sample code are likely way out of date. i don't know what Gitpod is so i guess the answer to that question is no!

**Comment by StephanLuis** [comment]
Hi It's taken some time to run my test on Appium 2.0, the 'the added 0' problem as described above still persists. Please have a look, the appium log is attached.  Thanks.
[appium.log](https://github.com/appium/appium/files/8910632/appium.log)

**Comment by StephanLuis** [comment]
@jlipps  @AutomatedTester could you **please** have another look at this?  It's really important to me to have automated testing and surprisingly, a pretty serious bug in appium.  I'll be happy to help in any way that I can.  Many thanks.

**Comment by jlipps** [comment]
@StephanLuis didn't we decide that the culprit must be your code that is adding 0s? What was the result of that investigation?

**Comment by StephanLuis** [comment]
@jlipps  no!  In every other test environment than iOS there is no bug in the return value (leading 0s).  Though I think we discussed a possibility that appium may be triggering additional events.  But that is just a guess ... and still a bug within appium.  The actual browser return value on iOS is fine.

**Comment by jlipps** [comment]
So all Appium is doing when you call `setValue` like you are in this script is:

1. Running the [click](https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/atoms/action.js#L290-L296) selenium atom on the element
2. Running the [type](https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/atoms/action.js#L166-L241) atom on the element

So if there is a bug, it would be somewhere in these Selenium atoms (which are not maintained by Appium; Appium just uses them). It's possible Appium's atoms are outdated and we could bring in newer atoms (maybe what you're experiencing is a bug which has been fixed?).

But I would recommend reading through this code as it will give you an idea of the JavaScript events which are fired on the field. I notice for example that there is a lot of special handling for mobile safari and date fields which results in a lot of events being fired. Maybe this is interacting poorly with your code's event handlers?

**Comment by StephanLuis** [comment]
I can had a look at those links.  Nothing really stands out.  Can you give a link to what executes for `driver.executeScript('return document.querySelector("#bob").value')  `   Thanks.

**Comment by jlipps** [comment]
https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/atoms/inject.js#L271-L283 is what happens when we call executeScript. basically it wraps the string as a function, and calls `.apply` on it. doesn't seem like any events would be fired here.

@KazuCocoa do you know how recent our selenium atoms are?

**Comment by KazuCocoa** [comment]
No, i tried to see the diff in https://github.com/appium/appium-remote-debugger/pull/258 to update atoms, but haven't deeply dug it yet.

**Comment by KazuCocoa** [comment]
I haven't tried, but potentially atoms update in https://github.com/appium/appium-remote-debugger/pull/258/files might help. (You can apply the change locally)

**Comment by StephanLuis** [comment]
@KazuCocoa @jlipps  Thanks for the link.  Who should apply the change?  I am running appium as a service (through SauceLabs) so cannot update.

**Comment by StephanLuis** [comment]
@jlipps @KazuCocoa What environment do you use? I don't see 'everything' on [contributing-to-appium](https://github.com/appium/appium/tree/master/docs/en/contributing-to-appium).  Wondering (aside from this bug) if there would be funding to set up a docker container with the 'certified' dev environment for Appium?  Then contributors could pick that up and work on a fork without any setup!  Let me know what you think? Could yield pretty quick results.

**Comment by StephanLuis** [comment]
@jlipps Since I've been using SAAS appium, I don't know much about how the setup and testing would work.  Am I right, each driver is tested using the sdk specific for each OS, ex android, iOS, ... ?

 Two quick questions about this side topic, would [Docker-OSX](https://github.com/sickcodes/Docker-OSX) help? Would an Appium dev docker env for browser /JS only (for various desktop and mobile OS) be achievable without the sdks?

**Comment by KazuCocoa** [comment]
The result of WebView depends on a real device or a simulator as well. If you used a real device and you did not have a real device on your local, perhaps the result could differ, unfortunately. I haven't caught up with Apple's licenses for macOS recently, so i cannot say the Docker image is helpful here, but it may help to use a simulator environment theoretically.

Basic requirements for each driver are:
https://github.com/appium/appium-uiautomator2-driver#requirements
https://github.com/appium/appium-xcuitest-driver#requirements

**Comment by StephanLuis** [comment]
@jlipps @KazuCocoa So... what do you think the best way to address this 007 bug would be??  BTW saucelabs is offering Appium 2 in a beta service.

**Comment by jlipps** [comment]
Unfortunately I think the next step is to try to get the latest atoms working in the appium remote debugger. I have started this process but the atoms have changed so much it is not straightforward. Some atoms have changed names, some have been removed entirely, and so on. Unfortunately I don't have time allotted to focus on this issue right now. But if someone wants, the results of my building the most recent atoms and hand copying/renaming some of them into our remote debugger is here: https://github.com/appium/appium-remote-debugger/pull/268

@StephanLuis as a workaround right now, why not just set the value with executeScript instead of sendKeys?

**Comment by StephanLuis** [comment]
@jlipps @KazuCocoa   I've posted another bug relating to this.  Appium select and click are activating a neighbouring, incorrect element.  [Click/Select Bug](https://github.com/appium/appium/issues/17415).  Could you please have a look?  Thanks.

**Comment by mykola-mokhnach** [comment]
@StephanLuis Would it be possible to prepare a test page where the described bug could be reliably reproduced (`a time input plugin control that have an extra 0 using appium for iOS.`)? Without reliable reproducibility steps steps we won't be able to address this issue

**Comment by StephanLuis** [comment]
This should give the controls to see the appium error.

https://stephanluis.github.io/input-duration/

**Comment by mykola-mokhnach** [comment]
Thanks for the repro page — that made it possible to actually verify this instead of guessing.

Traced it down to three separate bugs in the remote-debugger's type() atom (the code that simulates keystrokes for Safari sendKeys), all now fixed in appium/appium-remote-debugger#531:

1. A fast path for type="number" fields (added for a previous fix, #18765) was batching any numeric string into a single value assignment instead of dispatching real keystroke-by-keystroke events. Your widget's own auto-advance/validation logic listens on keyup, so it never ran at all — digits just piled up in the first field.
2. The synthetic keyboard events this atom fires never set the standard KeyboardEvent.key/.code properties, only the old keyCode/charCode. Your widget's filter checks event.key, so even once (1) was fixed, the filter still silently no-opped.
3. The simulated keyboard targeted a fixed element for the whole type() call, so when your widget's JS moved focus to the next field mid-typing, subsequent keystrokes kept landing on the old field instead of following focus — a real keyboard always types wherever focus currently is.

With all three fixed, sending "777777777" to your widget now correctly lands each digit in the right field: 77:7:7.777. The one thing I couldn't close is the leading-zero padding your widget applies via a change handler on blur (77:07:07.777 on a real browser) — browsers only fire change on blur for genuinely trusted user interaction, and this atom's typing is simulated via dispatchEvent, which doesn't set that flag. That's a platform-level limitation of DOM-injected keystroke simulation, not something fixable in this atom.

**Comment by mykola-mokhnach** [comment]
Closed as fixed. The newest driver should have the fix

**Comment by StephanLuis** [comment]
it would be great to have the leading zeros, what is the best we can do in
terms of keeping that aspect of the issue open?

Thanks,
Stephan

**Comment by mykola-mokhnach** [comment]
> it would be great to have the leading zeros, what is the best we can do in
> terms of keeping that aspect of the issue open?
> 
> Thanks,
> Stephan
> […](#)

Like I said it's not possible. Not because we don't want to, but because Safari does not allow it. There is no known workaround :(

**Comment by StephanLuis** [comment]
I love to be naive and persistent :) but gemini suggests firing a click on
a 'safe' element (possibly a new element injected for this purpose) and
then going back to the original element.  Do you think that work even
though it seems laborious to attach that to every dispatchEvent for
simulated typing, maybe a flag could turn this feature on only when the
browsers 'trusted user interaction' is causing a problem.

https://share.google/aimode/kJRJ5dxpPgYrRnwOC

Many thanks for your patience,

Stephan
